app/src/main/java/cm/aptoide/pt/AptoideApplication.java

```diff
@@ -344,6 +347,18 @@ public LeakTool getLeakTool() {
     analyticsManager.setup();
     invalidRefreshTokenLogoutManager.start();
     aptoideDownloadManager.start();
+
+    Futures.addCallback(getAsyncApplicationComponent().moPubAdsService(),
+        new FutureCallback<MoPubAdsService>() {
+          @Override public void onSuccess(@javax.annotation.Nullable MoPubAdsService service) {
+            if (service != null) {
+              service.observeWalletInstalation();
+            }
+          }
+
+          @Override public void onFailure(Throwable t) {
+          }
+        });
```

Comment:
I changed this completely so we don't compromise the architecture. The issue was that some of the MoPubAdsService dependencies depended on a Google Advertising ID fetch that needs to be retrieved in a background thread. Since regular injection runs on the main thread, this needed to be changed. The dependency tree of this service was also way too complex to initialize manually.

Even though I changed this so that the retrieval of the Google Advertising Id is now async inside the relevant dependency, it might be beneficial to explore asynchronous injection in the future (e.g to speed up launch time, which according to trace logs, the initial application injection setup takes about 400ms, not accounting for other dependency injections such as in activities, fragments...!).

Even though I used producers, there's more elegant ways to achieve async injection with RxJava explored here:
https://proandroiddev.com/dagger-2-part-three-new-possibilities-3daff12f7ebf